phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TaskRegionObserver.java

```diff
@@ -0,0 +1,261 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.coprocessor;
+
+import java.io.IOException;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.sql.Types;
+
+import java.util.Optional;
+import java.util.Properties;
+import java.util.TimerTask;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.CoprocessorEnvironment;
+import org.apache.hadoop.hbase.coprocessor.ObserverContext;
+import org.apache.hadoop.hbase.coprocessor.RegionCoprocessor;
+import org.apache.hadoop.hbase.coprocessor.RegionCoprocessorEnvironment;
+import org.apache.hadoop.hbase.coprocessor.RegionObserver;
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+import org.apache.phoenix.jdbc.PhoenixConnection;
+import org.apache.phoenix.jdbc.PhoenixDatabaseMetaData;
+import org.apache.phoenix.jdbc.PhoenixDriver;
+import org.apache.phoenix.query.QueryServices;
+import org.apache.phoenix.query.QueryServicesOptions;
+import org.apache.phoenix.schema.MetaDataClient;
+import org.apache.phoenix.schema.PTable;
+import org.apache.phoenix.schema.PTable.TaskType;
+
+import org.apache.phoenix.util.PhoenixRuntime;
+import org.apache.phoenix.util.QueryUtil;
+
+
+/**
+ * Coprocessor for task related operations. This coprocessor would only be registered
+ * to SYSTEM.TASK table
+ */
+@SuppressWarnings("deprecation")
+public class TaskRegionObserver implements RegionObserver, RegionCoprocessor {
+    public static final Log LOG = LogFactory.getLog(TaskRegionObserver.class);
+    protected ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(TaskType.values().length);
+    private long timeInterval = QueryServicesOptions.DEFAULT_TASK_HANDLING_INTERVAL;
+    private long timeMaxInterval = QueryServicesOptions.DEFAULT_TASK_HANDLING_MAX_INTERVAL;
+    @GuardedBy("TaskRegionObserver.class")
+    // Added for test purposes
+    private long initialDelay = 0;
+
+    @Override
+    public void preClose(final ObserverContext<RegionCoprocessorEnvironment> c,
+            boolean abortRequested) {
+        executor.shutdownNow();
+    }
+
+    @Override
+    public Optional<RegionObserver> getRegionObserver() {
+        return Optional.of(this);
+    }
+
+    @Override
+    public void start(CoprocessorEnvironment env) throws IOException {
+        Configuration config = env.getConfiguration();
+        timeInterval =
```

Comment:
nit: good to have comment or doc here about all the props that we are making configurable.

Reply:
We can actually simplify by not exposing these properties out. The defaults seem to be good enough.
Especially the `initialDelay` parameter which is only used for test purposes.

Reply:
@karanmehta93, I am new to the Phoenix code and following the styles in the existing code base. MetaDataObserver has these parameters defined for the same purposes. I can imagine that we may need to change these values if we hit some issues and a single default value may not be sufficient. I am not sure how you conclude that initialDelay is/will be only used for only test purposes.

Reply:
> I am not sure how you conclude that initialDelay is/will be only used for only test purposes.

This is something that's only applicable for once during the lifetime of a region server, since phoenix lies inside HBase process. I am not sure if this will help. This is not the delay that will be applied before every task, is that correct? Also, I see that initialDelay is commented as "for test purposes" on Line 72 in TaskRegionObserver.java.

Reply:
@karanmehta93, Ok, I will remove the configuration parameter for the initial delay